### ISSUE TITLE
Kwargs for player calls

### ISSUE THREAD

**Issue by yerkesoul** [clp-research/clemcore#39]
Could you please allow kwargs for player calls?

**Comment by davidschlangen** [comment]
Can you elaborate / give some examples? What should such a call look like, and what would you want to achieve with it?

**Comment by Gnurro** [comment]
Calling a Player instance (if not modified far beyond the base class) defaults to the generation call, which may or may not take more arguments depending on backend. I assume this is about passing more specific generation arguments like maximum number of generated tokens and the like?

**Comment by phisad** [comment]
This was about implementing a programmatic player that requires more information than simply the message history to generate a dynamic and accurate response (here in particular the positioning of the players in a gridworld). 

This could be a more general requirement: every time the Player need to make a generation attempt which requires more information than the message history. This is not possible for now. 

We should provide something like a `world_state` argument to the call (or better replace the simple `turn_idx` argument with that) as a holder of more structured information or information of other modalities e.g. vision.

This would allow to easier change the player backends from LLM, to human (slurk) to programmatic (a bot). This is not so easily possible for now, but the initial idea was to overwrite `__call__` entirely (which would not allow for a backend change easily).

**Comment by davidschlangen** [comment]
So the idea is that the player might need more information than what is in the context that is used for generation, or more easy access to that information? That makes sense. And I love the idea of abstracting this as the world state. (Which of course would bring things a bit more in line with reinforcement learning gym setups. In fact, why not allow for an optional reward argument as well?)

**Comment by phisad** [comment]
OK, I see now that this is also connected to the way how we might want to pass the temperature and max_tokens args.
I was experimenting with the idea to change generate response to:

`def generate_response(self, messages: List[Dict], gen_args: GenerateArguments) -> Tuple[Any, Any, str]:``

where we have GenerateArguments (similar to ModelSpec)

```
class GenerateArguments(SimpleNamespace):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

    def __repr__(self):
        return f"GenerateOptions({str(self)})"

    def __str__(self):
        return str(self.__dict__)

    def __getitem__(self, item):
        """ dict-like behavior """
        return getattr(self, item)

    def __contains__(self, item):
        """ dict-like behavior """
        return self.has_attr(item)

    def has_attr(self, attribute):
        return hasattr(self, attribute)
```

The problem with this approach is that we need to pass the temperature from the `Benchmark` through the `GameMaster` to the `Player` which requires additional changes in the games. 

Thus, I will for now add these generation arguments to the `Model` directly (as these are already early initialised and already given to the Players). (I guess, to actually properly design this we would actually need a `Decoder/Generator` class.)

I can then still add the world_state argument to generate_response.

**Comment by davidschlangen** [comment]
Can this be closed, @phisad ?

**Comment by phisad** [comment]
Closed because duplicate of #56
